crypto/dilithium/p_dilithium3_asn1.c

```diff
@@ -0,0 +1,208 @@
+  if (key == NULL) {
+    OPENSSL_PUT_ERROR(EVP, ERR_R_MALLOC_FAILURE);
+    return 0;
+  }
+
+  OPENSSL_memcpy(key->priv, in, len);
+  key->has_private = 1;
+
+  dilithium3_free(pkey);
+  pkey->pkey.ptr = key;
+  return 1;
+}
+
+static int dilithium3_set_pub_raw(EVP_PKEY *pkey, const uint8_t *in, size_t len) {
+  if (len != DILITHIUM3_PUBLIC_KEY_BYTES) {
+    OPENSSL_PUT_ERROR(EVP, EVP_R_INVALID_KEYBITS);
+    return 0;
+  }
+
+  DILITHIUM3_KEY *key = OPENSSL_malloc(sizeof(DILITHIUM3_KEY));
+  if (key == NULL) {
+    OPENSSL_PUT_ERROR(EVP, ERR_R_MALLOC_FAILURE);
+    return 0;
+  }
+
+  OPENSSL_memcpy(key->pub, in, len);
+  key->has_private = 0;
+
+  dilithium3_free(pkey);
+  pkey->pkey.ptr = key;
+  return 1;
+}
+
+static int dilithium3_get_priv_raw(const EVP_PKEY *pkey, uint8_t *out,
+                                   size_t *out_len) {
+  const DILITHIUM3_KEY *key = pkey->pkey.ptr;
+  if (!key->has_private) {
+    OPENSSL_PUT_ERROR(EVP, EVP_R_NOT_A_PRIVATE_KEY);
+    return 0;
+  }
+
+  if (out == NULL) {
+    *out_len = DILITHIUM3_PRIVATE_KEY_BYTES;
+    return 1;
+  }
+
+  if (*out_len < DILITHIUM3_PRIVATE_KEY_BYTES) {
+    OPENSSL_PUT_ERROR(EVP, EVP_R_BUFFER_TOO_SMALL);
+    return 0;
+  }
+
+  OPENSSL_memcpy(out, key->priv, DILITHIUM3_PRIVATE_KEY_BYTES);
+  *out_len = DILITHIUM3_PRIVATE_KEY_BYTES;
+  return 1;
+}
+
+static int dilithium3_get_pub_raw(const EVP_PKEY *pkey, uint8_t *out,
+                                  size_t *out_len) {
+  const DILITHIUM3_KEY *key = pkey->pkey.ptr;
+  if (out == NULL) {
+    *out_len = DILITHIUM3_PUBLIC_KEY_BYTES;
+    return 1;
+  }
+
+  if (*out_len < DILITHIUM3_PUBLIC_KEY_BYTES) {
+    OPENSSL_PUT_ERROR(EVP, EVP_R_BUFFER_TOO_SMALL);
+    return 0;
+  }
+
+  OPENSSL_memcpy(out, key->pub, DILITHIUM3_PUBLIC_KEY_BYTES);
+  *out_len = DILITHIUM3_PUBLIC_KEY_BYTES;
+  return 1;
+}
+
+static int dilithium3_pub_decode(EVP_PKEY *out, CBS *params, CBS *key) {
+  // See https://datatracker.ietf.org/doc/draft-ietf-lamps-dilithium-certificates/ section 4.
+  // The parameters must be omitted.
+  if (CBS_len(params) != 0) {
+    OPENSSL_PUT_ERROR(EVP, EVP_R_DECODE_ERROR);
+    return 0;
+  }
+
+  return dilithium3_set_pub_raw(out, CBS_data(key), CBS_len(key));
+}
+
+static int dilithium3_pub_encode(CBB *out, const EVP_PKEY *pkey) {
+  const DILITHIUM3_KEY *key = pkey->pkey.ptr;
+
+  // See https://datatracker.ietf.org/doc/draft-ietf-lamps-dilithium-certificates/ section 4.
+  // TODO: finalize this definition - OCTETSTRING to BITSTRING conversion.
+  CBB spki, algorithm, oid, key_bitstring;
+  if (!CBB_add_asn1(out, &spki, CBS_ASN1_SEQUENCE) ||
+      !CBB_add_asn1(&spki, &algorithm, CBS_ASN1_SEQUENCE) ||
+      !CBB_add_asn1(&algorithm, &oid, CBS_ASN1_OBJECT) ||
+      !CBB_add_bytes(&oid, dilithium3_asn1_meth.oid, dilithium3_asn1_meth.oid_len) ||
+      !CBB_add_asn1(&spki, &key_bitstring, CBS_ASN1_BITSTRING) ||
+      !CBB_add_u8(&key_bitstring, 0 /* padding */) ||
```

Comment:
@[andrewhop](https://github.com/andrewhop) andrewhop [on Oct 4](https://github.com/awslabs/aws-lc/pull/637/files#r987427443)
What is this padding for?

Reply:
Still working details out like this in the draft spec, but essentially, the padding is required to make the bit string total size to 15616 bit -> 1952 bytes as per the [Dilithium spec](https://pq-crystals.org/dilithium/).